server/src/utils/mixerConnections/DHDConnection.ts

```diff
@@ -0,0 +1,861 @@
+  method: "update"
+  payload: any
+}
+
+interface DHDEventMessage {
+  method: "event"
+  payload: unknown
+  type: unknown
+}
+
+interface DHDUnsubscribeReqMessage extends DHDMessageBase {
+  method: "unsubscribe"
+  path: string
+}
+
+interface DHDUnsubscribeResMessage extends DHDMessageBase {
+  method: "unsubscribe"
+  path: string
+  success: true
+}
+
+type DHDAnyResMessage =
+  | DHDErrResMessage
+  | DHDUnsubscribeResMessage
+  | DHDSubscribeResMessage
+  | DHDAuthResMessage
+  | DHDGetResMessage
+  | DHDSetResMessage
+
+type DHDIncomingMessage = DHDAnyResMessage | DHDUpdateMessage
+type DHDOutgoingMessage = DHDAuthReqMessage | DHDSetReqMessage | DHDGetReqMessage | DHDSubscribeReqMessage | DHDUnsubscribeReqMessage
+type DHDUntaggedOutgoingMessage = Omit<DHDOutgoingMessage, 'msgID'>
+
+type DHDResponseHandler = (msg: Readonly<DHDAnyResMessage>) => void
+type DHDUpdateHandler<T> = (subTreeValue: Readonly<T>) => void
+
+class DHDWebSocketClient extends EventEmitter<{
+  'error': [string]
+  'warn': [string]
+  'close': [],
+  'open': [],
+  'event': [DHDEventMessage["type"], DHDEventMessage["payload"]]
+}> {
+
+  private msgIDListeners: Map<number, DHDResponseHandler> = new Map()
+  private updateListeners: Map<string, DHDUpdateHandler<unknown>[]> = new Map()
+
+  private protocolLastMsgID = 0
+
+  private wsConnection: WebSocket
+
+  private pingInterval: NodeJS.Timeout
+
+  constructor(url: string, token: string, private readonly options?: {
+    pingInterval?: number
+  }) {
+    super()
+
+    this.setupConnection(url, token)
+  }
+
+  private setupConnection = (url: string, token: string) => {
+    this.wsConnection = new WebSocket(url)
+
+    this.wsConnection.addListener('error', (error: any) => {
+      if (
+        (error.message + '').match(/econnrefused/i) ||
+        (error.message + '').match(/disconnected/i)
+      ) {
+        this.emit('error', `WebSockets connection not established: ${error.message}`)
+      } else {
+        this.emit('error', `WebSockets connection unknown error: ${error}`)
+      }
+    })
+    this.wsConnection.addListener('message', this.onMessage)
+    this.wsConnection.addListener('close', () => {
+      this.emit('close')
+      if (this.pingInterval !== undefined) {
+        clearInterval(this.pingInterval)
+      }
+    })
+    this.wsConnection.addListener('open', () => {
+      this.authorize(token).then(() => {
+        if (this.options?.pingInterval) {
+          this.pingInterval = setInterval(() => {
+            this.wsConnection.ping()
+          }, this.options.pingInterval)
+        }
+        this.emit('open')
+      }).catch((err) => {
+        this.emit('error', `Could not authorize: ${err}`)
+        this.close()
+      })
+    })
+  }
+
+  private getNextMsgID = (): number => {
+    const msgID = this.protocolLastMsgID++
+    // wrap around the msgID when needed
+    if (this.protocolLastMsgID >= Number.MAX_SAFE_INTEGER) {
+      this.protocolLastMsgID = Number.MIN_SAFE_INTEGER
```

Comment:
Nit-picking:
`Number.MAX_SAFE_INTEGER` is safe in javascript/v8/nodejs, I wouldn't assume that that is safe on the dhd-side.

```suggestion
    if (this.protocolLastMsgID >= 2147483647) { // 2^31 - 1
      this.protocolLastMsgID = 0
```

Reply:
The documentation says that the `msgID` can be "anything", and that whatever is sent here will be sent back, so the roll-over is for our sake only. That's why i chose `Number.MAX_SAFE_INTEGER` and `Number.MIN_SAFE_INTEGER`.